src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/FormFileModelBinder.cs

```diff
@@ -36,66 +36,151 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
         {
-            // If we're at the top level, then use the FieldName (paramter or property name).
-            // This handles the fact that there will be nothing in the ValueProviders for this parameter
-            // and so we'll do the right thing even though we 'fell-back' to the empty prefix.
-            var modelName = bindingContext.IsTopLevelObject
-                ? bindingContext.BinderModelName ?? bindingContext.FieldName
-                : bindingContext.ModelName;
 
-            object value;
+            ICollection<IFormFile> collection;
             if (bindingContext.ModelType == typeof(IFormFile))
             {
-                var postedFiles = await GetFormFilesAsync(modelName, bindingContext);
-                value = postedFiles.FirstOrDefault();
+                // Create an intermediate list. Will throw it away later.
+                collection = new List<IFormFile>();
             }
-            else if (typeof(IEnumerable<IFormFile>).IsAssignableFrom(bindingContext.ModelType))
+            else if (bindingContext.ModelType == typeof(IFormFile[]))
             {
-                var postedFiles = await GetFormFilesAsync(modelName, bindingContext);
-                value = ModelBindingHelper.ConvertValuesToCollectionType(bindingContext.ModelType, postedFiles);
+                if (bindingContext.ModelMetadata.IsReadOnly)
+                {
+                    // Can't change the length of an existing array or replace it.
+                    return null;
+                }
+
+                // Use a List<IFormFile> for now. Will create an array later.
+                collection = new List<IFormFile>();
             }
             else
             {
-                // This binder does not support the requested type.
-                Debug.Fail("We shouldn't be called without a matching type.");
-                return;
+                if (bindingContext.Model == null)
+                {
+                    if (bindingContext.ModelMetadata.IsReadOnly)
+                    {
+                        // Need a new collection instance but unable to assign it to the property.
+                        return null;
+                    }
+
+                    if (bindingContext.ModelType == typeof(IFormFileCollection))
+                    {
+                        // Special-case creating a custom IFormFileCollection if the property is settable and
+                        // currently null. Use a List<IFormFile> for now. Will create a FileCollection later.
+                        collection = new List<IFormFile>();
+                    }
+                    else
+                    {
+                        // Note this call may return null if the model type cannot be activated.
+                        collection = ModelBindingHelper.CreateCompatibleCollection<IFormFile>(
+                            bindingContext.ModelType,
+                            capacity: null);
```

Comment:
Couple of thoughts
- The read-only `ICollection<T>` stuff is definitely something I'd like to see supported, consistent w/ other collection properties in the system.
- Less sure about the `try` block around activating the user's type. But that's fairly localized and not repeated.
- Could simplify things a bit if we were more consistent about read-only scalars. The code here and in `CollectionModelBinder` assumes `MutableObjectModelBinder` or `TryUpdateModelAsync()` has already excluded the read-only scalar case. But it complicates the code to avoid checking `IsReadOnly` for the scalar case. Bottom line we could centralize some of this in a `Helper<T>` that checks
```c#
bindingContext.ModelMetadata.IsReadOnly &&
  (bindingContext.Model == null && !typeof(ICollection<T>).IsAssignableFrom(bindingContext.ModelType) ||
   bindingContext.ModelType = typeof(T[]))
```
- `ModelBindingHelper.CreateCompatibleCollection<T>()` could take a `ModelBindingContext` parameter instead of passing `bindingCotnext.ModelType`. With the above out of the way, this helper could handle the bare `T` and `T[]` cases.
- That would simplify this block to something like
``` c#
if (CanReuseOrCreateModel(bindingContext))
{
    return null;
}

ICollection<IFormFile> collection;
if (bindingContext.Model == null && bindingContext.ModelType == typeof(IFormFileCollection))
{
    collection = new List<IFormFile>();
}
else
{
    collection = CreateCompatibleCollection<IFormFile>(ModelType, capacity: null);
}
```
- Can't combine further w/o pushing the `IFormFileCollection` stuff down where it doesn't belong.

Thoughts?